third_party/nginx_latency_status_module/src/ngx_http_latency_recorder.c

```diff
@@ -0,0 +1,123 @@
+// Copyright (C) 2002-2016 Igor Sysoev
+// Copyright (C) 2011-2016 Nginx, Inc.
+// Copyright (C) 2020 Google Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+// 1. Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+// 2. Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+// SUCH DAMAGE.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "ngx_http_latency_recorder.h"
+#include "ngx_http_latency_storage.h"
+#include "ngx_http_latency_stub_status_module.h"
+
+
+// Get the index of the distribution bucket that a latency fall in.
+ngx_int_t get_latency_index(ngx_int_t max_exponent, ngx_int_t *latency_bucket_bounds, ngx_int_t latency) {
+  for (ngx_int_t i = 0; i <= max_exponent; i++) {
+    if (latency < latency_bucket_bounds[i]) {
+      return i;
+    }
+  }
+
+  // return the index for the overflow bucket.
+  return max_exponent + 1;
+}
+
+
+// Update a latency record with a single new latency.
+void update_latency_record(latency_stat *record, ngx_int_t latency, ngx_http_latency_main_conf_t *main_conf) {
+  ngx_int_t distribution_index;
+
+  ngx_atomic_fetch_add(&record->request_count, 1);
+  ngx_atomic_fetch_add(&record->sum, latency);
+  ngx_atomic_fetch_add(&record->sum_squares, latency * latency);
+
+  distribution_index = get_latency_index(main_conf->max_exponent, main_conf->latency_bucket_bounds, latency);
+  ngx_atomic_fetch_add(&record->distribution[distribution_index], 1);
+}
+
+
+// Record the latency statistics.
+ngx_int_t ngx_http_record_latency(ngx_http_request_t *r)
+{
+  ngx_time_t *now;
+  ngx_msec_int_t request_time;
+  ngx_http_latency_conf_t *conf;
+  ngx_http_latency_main_conf_t *main_conf;
+  ngx_http_latency_shm_t *latency_record;
```

Comment:
In accordance to the C++ style guide (https://engdoc.corp.google.com/eng/doc/devguide/cpp/styleguide.md?cl=head#Local_Variables), could you initialize variables in the declaration? There are a few other places in this file and other files as well.

Reply:
done